File: gcs.py
# This code is released into the public domain
from __future__ import division
import sys, math
from hashlib import md5
import struct
from array import array

def bitwriter(f):
    """
    Coroutine to write bits into a file-like.
    Use .send((N, V)) to write the N least-significant
    bits of V into the file.
    Remember to call .close() to flush the internal
    state when done.
    """
    v = 0
    n = 0
    try:
        while 1:
            (n2,v2) = yield None
            v <<= n2
            v |= v2 & ((1<<n2)-1)
            n += n2
            while n >= 8:
                b = (v >> (n-8)) & 255
                f.write(chr(b))
                n -= 8
            v &= 255
    except GeneratorExit:
        if n != 0:
            v &= (1 << n) - 1
            f.write(chr(v))
        raise

def bitreader(f):
    """
    Coroutine to read bits from a file-like.
    Use .send(N) to read N bits from the file.
    """
    accum = 0
    n = 0
    v = None
    while 1:
        n2 = yield v
        while n <= n2:
            accum <<= 8
            try:
                accum |= ord(f.read(1))
            except TypeError:
                # ord(None) => eof
                return
            n += 8
        v = (accum >> (n-n2)) & ((1<<n2)-1)
        n -= n2
        accum &= (1<<n)-1


def golomb_enc(f, P):
    """
    Coroutine to encode Golomb/Rice-encoded values into
    a file-like f, with parameter P.
    Use .send(V) to encode the value V into the file.
    This is an optimal encoding if values are geometrically
    distributed with P.
    """
    logp = int(math.log(P,2))
    f = bitwriter(f); f.next();
    while 1:
        v = yield None
        q,r = v//P, v%P
        f.send((q+1, (1<<(q+1))-2))
        f.send((logp, r))   

def golomb_dec(f, P):
    """
    Coroutine to decode Golomb/Rice-encoded values
    from a file-like f, with parameter P.
    Use .next() (or iterate over it) to read the values.
    """
    logp = int(math.log(P,2))
    f = bitreader(f); f.next()
    while 1:
        try:
            v = 0
            while f.send(1):
                v += P
            v += f.send(logp)
        except StopIteration:
            return
        yield v

def gcs_hash(w, (N,P)):
    """
    Hash value for a GCS with N elements and 1/P probability
    of false positives.
    We just need a hash that generates uniformally-distributed
    values for best results, so any crypto hash is fine. We
    default to MD5.
    """
    h = md5(w).hexdigest()
    h = long(h,16)
    return h % (N*P)

class GCSBuilder:
    def __init__(self, N, P):
        self.N = N
        self.P = P
        self.values = array("L")
        self.values.append(0)
        
    def add(self, v):
        self.values.append(gcs_hash(v, (self.N,self.P)))

    def finalize(self, f):
        values = sorted(self.values)
        f.write(struct.pack("=LL", self.N, self.P))
        f = golomb_enc(f, self.P)
        f.next()
        for i in range(len(values)-1):
            d = values[i+1] - values[i]
            if d == 0:
                continue
            f.send(d)
        f.close()

class GCSQuery:
    def __init__(self, f):
        self.f = f
        self.N, self.P = struct.unpack("=LL", self.f.read(8))

    def _rewind(self):
        self.f.seek(8)

    def query(self, w):
        h = gcs_hash(w, (self.N,self.P))
        N = 0
        self._rewind()
        for d in golomb_dec(self.f, self.P):
            N += d
            if h == N:
                return True
            if h < N:
                # early-exit since values are sorted
                return False
        return False
            

if __name__ == "__main__":
    if sys.argv[1] == "build":
        prob = 2**10
        words = open(sys.argv[2]).readlines()
        gcs = GCSBuilder(len(words), prob)
        for w in words:
            gcs.add(w.strip())
        with open("table.gcs", "wb") as f:
            gcs.finalize(f)
            fsize = f.tell()
        print "Number of words: %d" % len(words)
        print "False positives: %f" % (1/prob)
        print "Size: %d" % (fsize)
        print "Bits per word: %f" % (fsize*8 / len(words))

    elif sys.argv[1] == "query":
        gcs = GCSQuery(open("table.gcs","rb"))
        for w in sys.argv[2:]:
            print gcs.query(w)

File: js/gcs.js
function bitreader(arr) {
  var offset = 0,
      accum = 0,
      n = 0;

  function c (n2, v) {
    if (typeof v === 'undefined') v = 0;
    if (n2 > 8) {
      v = v * 256 + c(8);
      return c(n2-8, v);
    } else {
      n -= n2;
      if (n < 0) {
        if (offset >= arr.length) throw "End of array";
        accum = (accum << 8) | arr[offset++];
        n += 8;
      }
      v = v * Math.pow(2, n2) + (accum >>> n);
      accum &= (1 << n) - 1;
      return v;
    }
  }
  return c;
}

function bitwriter(arr) {
  var accum = 0,
      n = 0,
      tmp = 0;
  function c (n2, v2) {
    if (n2 > 8) {
      n2 -= 8;
      tmp = v2 / Math.pow(2, n2) >>> 0;
      c(8, tmp);
      c(n2, v2 - tmp * Math.pow(2, n2));
    } else {
      accum = accum * Math.pow(2, n2) + v2;
      n += n2;
      if (n >= 8) {
        arr.push(accum / Math.pow(2, n-8));
        n -= 8;
        accum = accum & ((1 << n) - 1);
      }
    }
  }

  c.close = function () {
    if (n !== 0) {
      accum = (accum << (8-n)) & 255;
      arr.push(accum);
    }
  };

  return c;
}

function gcs_hash(w, N, P) {
    h = md5(w);
    h = parseInt(h.substring(24,32), 16) % (N*P);
    return h;
}

function golomb_enc(arr, P) {
  var logp = Math.round(Math.log(P) * Math.LOG2E);
  var f = bitwriter(arr);
  function c (v) {
    var q = ~~(v / P),
        r = v % P;
    f(q+1, (1 << (q+1)) - 2);
    f(logp, r);
    return;
  }

  c.close = function () {
    f.close();
  };

  c.write = f;

  return c;
}

function golomb_dec(arr, P) {
  var logp = Math.round(Math.log(P) * Math.LOG2E);
  var f = bitreader(arr);
  var v;
  return function () {
    while(1) {
      v = 0;
      while (f(1)) {
        v += P;
      }
      var tmp = f(logp);
      v += tmp;
      return v;
    }
  };
}

function GCSBuilder(_N, _P) {
  var N = _N, P = _P, values = [0], words = [];

  this.add = function (v) {
    words.push(v);
    values.push(gcs_hash(v, N, P));
  };

  this.finalize = function () {
    var i,
        d,
        ab,
        res = [],
        header = new Array(8),
        f = golomb_enc(res, P);
    values.sort(function (a, b) { return a - b; });
    for (i = 0; i < values.length - 1; i += 1) {
      d = values[i+1] - values[i];
      if (d === 0 && i > 0) {
        continue;
      }
      f(d);
    }
    f.close();
    res = header.concat(res);
    res = new Uint8Array(res);
    dw = new DataView(res.buffer);
    dw.setUint32(0, N);
    dw.setUint32(4, P);
    return res.buffer;
  };
}

function GCSQuery(_arrBuff) {
  var dw = new DataView(_arrBuff),
    N = dw.getUint32(0),
    P = dw.getUint32(4),
    u8arr = new Uint8Array(_arrBuff, 8);
  this.query = function (w) {
    var h = gcs_hash(w, N, P),
      n = 0,
      d,
      f = golomb_dec(u8arr, P);

    while (1) {
      try {
        d = f();
        n += d;
        if (h === n) {
          return true;
        }
        if (h < n) {
          return false;
        }
      } catch (err) {
        break;
      }
    }
    return false;
  };
}